Chess/game.cpp

```diff
@@ -1,40 +1,166 @@
 #include "game.h"
+#include "desk.h"
+#include "command.h"
+
 #include <qfiledialog.h>
+#include <QVariant>
 
 Game::Game(QObject *parent) :
-    //QObject(parent),
-    viewer_(parent)
+    root_(parent),
+    desk_(NULL),
+    command_(NULL)
 {
     desk_ = new Desk(this);
 }
 
+Cell Game::parseQMLCellName(QString name)
+{
+    //needs err handle here
+    if (!name.contains(QRegExp("cell[1-8][1-8]")))
+        return Cell(-1,-1);
+
+
+    QString tmp = name.remove("cell");
+    int row = QString(tmp[0]).toInt()-1;
+    int col = QString(tmp[1]).toInt()-1;
+
+    return Cell(row,col);
+}
+
 void Game::cellAction(QString cell_name)
 {
+    Cell cell = parseQMLCellName(cell_name);
+    //bad input
+    if (cell == Cell(-1,-1)) {
+        interruptCommand();
+        return;
+    }
+
+    QObject *t_cell = root_->findChild<QObject*>(QString("t_")+cell_name);
+    //needs err handle here
+    if (t_cell == NULL) {
+        interruptCommand();
+        return;
+    }
+
+    if (command_ == NULL) {
+//        Figure *fig = desk_->getFigure(cell);
+//        if (fig == NULL)
+//            return;
 
+        if (desk_->getFigure(cell) == NULL)
+            return;
+
+
+        t_cell->setProperty("color", "red");
+
+        command_ = new Command(this, desk_);
+        command_->set_b_info(cell);
+    } else {
+        //Figure *fig = desk_->getFigure(cell);
+        command_->set_e_info(cell);
+
+        if (command_->valid()) {
+            command_->exec();
+            drawCommand();
+            // add to command list (for saving)
+
+            delete command_;
+            command_ = NULL;
+
+        } else {
+
+            interruptCommand();
+        }
+    }
 }
 
 void Game::startAction()
 {
+    interruptCommand();
+    desk_->clear();
     desk_->fillDefault();
+
+    drawCurState();
 }
 
 void Game::stopAction()
 {
-
+    interruptCommand();
+    desk_->clear();
+    drawCurState();
 }
 
 void Game::saveAction()
 {
-
+    interruptCommand();
 }
 
 void Game::loadAction()
 {
+    interruptCommand();
 //    QString filename = QFileDialog::getOpenFileName(
 //                this,
 //                tr("dsdasda"),
 //                "D://"
-//                );
+    //                );
+}
+
+void Game::tmpDraw()
+{
+    interruptCommand();
+    drawCurState();
+}
+
+void Game::drawCurState()
+{
+    for (size_t row=0; row<8; row++) {
+        for (size_t col=0; col<8; col++) {
+            QObject *t_cell = root_->findChild<QObject*>(QString("t_cell")+QString::number(row+1)+QString::number(col+1));

```

Comment:
you are translating indexes to string? do it only for output or saving, newer for processing. In case of processing numeric value is better